Update README.md

chrisb13 opened the pull request:
Clarifying definitions for different kinds of reproducibility checks [as agreed](https://access-nri.zulipchat.com/#narrow/channel/470339-model-dev/topic/Reproducibility/near/503147323)

dougiesquire commented on README.md:
```diff
@@ -207,7 +216,7 @@ The configuration properties needed to run the tests are:
 Pytest markers select what pytests to run in `model-config-tests`. Current markers include:
 
 - `checksum`: Historical reproducibility test that runs a model and compares checksums with a stored previous result.
-- `checksum_slow`: Restart reproducibility tests. These include model determinism, by repeating running a configuration from initial conditions to check it reproduces the same output, and checking reproducibility by confirming two short consecutive model runs give the same result as a longer single model run.
+- `checksum_slow`: Restart reproducibility tests. These include model determinism, by repeating running a configuration from initial conditions to check it reproduces the same output (`Determinism rest`--see above `Definitions for reproducibility`), and checking reproducibility by confirming two short consecutive model runs give the same result as a longer single model run (`Restart reproducibili ...
```
This is not actually correct anymore. Apologies, I should have updated as part of https://github.com/ACCESS-NRI/model-config-tests/pull/137. Do you mind if I push an update to this branch?

dougiesquire replied:
Done in [a8b5423](https://github.com/ACCESS-NRI/model-config-tests/pull/145/commits/a8b5423132f19b6073a87af09207b481d1566b3b)

dougiesquire commented on README.md:
```diff
@@ -6,6 +6,15 @@ The checksum pytests are used for reproducibility CI checks in this repository.
 
 Code from these pytests is adapted from COSIMAS's ACCESS-OM2's [bit reproducibility tests](https://github.com/COSIMA/access-om2/blob/master/test/test_bit_reproducibility.py).
 
+## Definitions for reproducibility
+It is helpful if we work from the same definitions of what 'reproducibility' means, we consider four kinds:
+ 1. _Determinism_.
+      1. _Determinism rest_: Repeated runs give the same result from rest;
+      1. _Determinism restart_: Repeated runs give the same result from a restart file;
+ 1. _Restart reproducibility_: Two short runs match a single longer run, in other words, reproducibility across a restart boundary;
+ 1. _Historical reproducibility_: Match a stored previous result;
+ 2. _Build reproducibilty_: Match a stored previous result when using the same build description files on the same machine.
+
```
I wonder it this should go at the end rather than right at the start? These definitions are relevant to some of the tests in this repo, but there are also lots of tests that have nothing to do with reproducibility.

chrisb13 replied:
Yeah sure.

But the new text added to [a8b5423](https://github.com/ACCESS-NRI/model-config-tests/pull/145/commits/a8b5423132f19b6073a87af09207b481d1566b3b) does not explicitly use these definitions so can you please update it (the marker descriptions I mean). I don't wish to change the above definitions (unless absolutely necessary) as they were agreed on by everyone at the cross-team meeting.

dougiesquire replied:
I don't understand sorry. The new marker descriptions use the exact same terminology as the definitions: historical reproducibility, restart reproducibility and determinism. Where is the ambiguity?

chrisb13 replied:
Okay, I just meant that the strings were very slightly different. 

I had a chat to @jo-basevi about this and in an ideal world they would match everywhere (example [1](https://github.com/ACCESS-NRI/access-om3-configs/pull/219#issuecomment-2787905313) and [2](https://github.com/search?q=repo%3AACCESS-NRI%2Fmodel-config-tests%20repro_historical&type=code)) but I think that's more work than it's worth.

Instead, I've updated [7b20f9b](https://github.com/ACCESS-NRI/model-config-tests/pull/145/commits/7b20f9b012ff1fd7fd15289b1de6dbac35db6b5d). You can see that I moved as discussed with Jo and some code blocks/relevant links.

Jo can now add her suggested changes on top.

dougiesquire commented on README.md:
```diff
@@ -113,7 +122,7 @@ Currently, these repositories make use of the reusable CI:
 - [access-esm1.6-configs](https://github.com/ACCESS-NRI/access-esm1.6-configs)
 - [access-om3-configs](https://github.com/ACCESS-NRI/access-om3-configs)
-- [access-om3-wav-configs](https://github.com/ACCESS-NRI/access-om3-wav-configs)
+- [access-om3-wav-configs](https://github.com/ACCESS-NRI/access-om3-wav-configs) (now depreciated with configs moved to `access-om3-configs`)
```
Just remove?

chrisb13 replied:
Yeah sure. I kept it in conservatively but don't mind.

dougiesquire replied:
Done in [2d6ff51](https://github.com/ACCESS-NRI/model-config-tests/pull/145/commits/2d6ff51ec7c56b76cc70be2d3ba87a02a7358536)

jo-basevi commented:
Added changes look good to me. 

The readme has a few examples referencing the old markers e.g. in the initial section for running `model-config-tests` manually on gadi and the example file for `config/ci.json`. I also wonder it's worth updating the markers descriptions (in `conftest.py`) to use the same terminology to those added in [a8b5423](https://github.com/ACCESS-NRI/model-config-tests/pull/145/commits/a8b5423132f19b6073a87af09207b481d1566b3b). These marker descriptions can be displayed with `model-config-tests --markers`
https://github.com/ACCESS-NRI/model-config-tests/blob/6246f28c56174c9ca770586f2cf78306102e4375/src/model_config_tests/conftest.py#L118-L133

I'm happy to add these updates to this PR, if that would be helpful?

chrisb13 commented on README.md:
```diff
@@ -6,6 +6,15 @@ The checksum pytests are used for reproducibility CI checks in this repository.
 
+## Definitions for reproducibility
+It is helpful if we work from the same definitions of what 'reproducibility' means, we consider four kinds:
+ 1. _Determinism_.
+      1. _Determinism rest_: Repeated runs give the same result from rest;
+      1. _Determinism restart_: Repeated runs give the same result from a restart file;
```
Thanks for the suggestion @aidanheerdegen, I'll add it.

> They're just using different inputs. Yes the code paths are different, but that's often the case for different configurations too.

The point for me is that in the ocean context, one could pass and the other might not and we tend to assume that if we have `Determinism restart` then we'll have `Determinism rest` but it's not necessarily true. By being explicit we are reminding ourselves that we shouldn't make this assumption.

chrisb13 commented:
Thanks @jo-basevi, as we discussed, that sounds great.

I've updated [7b20f9b](https://github.com/ACCESS-NRI/model-config-tests/pull/145/commits/7b20f9b012ff1fd7fd15289b1de6dbac35db6b5d) with the changes we discussed, [addressing @dougiesquire's feedback](https://github.com/ACCESS-NRI/model-config-tests/pull/145#discussion_r2069883764).

Feel free to add your changes on top.

jo-basevi commented:
Thanks, I've added the `README.md` changes. I moved the current pytest marker definitions to earlier in the document to the pytests specific section, and referenced it later on. I thought it might make more sense to group the `model-config-tests` pytest logic together.

@chrisb13 re chat yesterday on renaming test names and markers: We can easily change the name of tests to match the repro definitions without requiring PRs on other repositories, e.g. change `test_bit_repro_repeat` -> `test_repro_determinism_rest`. The markers are what is used in configuration files on model-configuration repositories to select tests used in CI/CD. Currently, the repro markers being used in CI/CD are `repro_historical` and `repro_restart` which I feel already align closely with the repro definitions.  This does mean that it would be easy to change the markers:
- `repro_restart_repeat` -> `repro_determinism_restart`
- `repro_repeat` -> `repro_determinism_rest`

Would that make more sense? If so, I can open a separate issue and PR.

chrisb13 commented:
Thanks for thinking this through @jo-basevi 

> Thanks, I've added the README.md changes. I moved the current pytest marker definitions to earlier in the document to the pytests specific section, and referenced it later on. I thought it might make more sense to group the model-config-tests pytest logic together.

Sounds great

>@chrisb13 re chat yesterday on renaming test names and markers: We can easily change the name of tests to match the repro definitions without requiring PRs on other repositories, e.g. change test_bit_repro_repeat -> test_repro_determinism_rest. The markers are what is used in configuration files on model-configuration repositories to select tests used in CI/CD. Currently, the repro markers being used in CI/CD are repro_historical and repro_restart which I feel already align closely with the repro definitions. This does mean that it would be easy to change the markers:

>repro_restart_repeat -> repro_determinism_restart
>repro_repeat -> repro_determinism_rest
> Would that make more sense? If so, I can open a separate issue and PR.

This sounds good in theory but I think `repro_restart_repeat` is not `determinism` it's the reproducibility one (across a restart boundary). Whereas this makes it nicely obvious: `repro_repeat -> repro_determinism_rest` but we best check in with @dougiesquire @minghangli-uni

jo-basevi commented:
After chatting with @chrisb13 and @dougiesquire, we decided that the following pytest markers be renamed:
- `repro_repeat` -> `repro_determinism` (not `repro_determinism_rest` as the configuration being tested may be starting from a restart, e.g. ACCESS-ESM1.6's [dev-preindustrial+concentrations config](https://github.com/ACCESS-NRI/access-esm1.6-configs/blob/2a33443c4a1c50f3f2b175fd7190bd2cf6fe8338/config.yaml#L122))
- `repro_restart_repeat` -> `repro_determinism_restart`

Note `repro_historical` and `repro_restart` markers will stay the same.

Test names (in `test_bit_reproducibility.py`) can be changed to be more consistent with the specific repro marker names, e.g.:
- `test_bit_repro_historical` -> `test_repro_historical`
- `test_bit_repro_repeat` -> `test_repro_determinism`
- `test_restart_repro` -> `test_repro_restart`
- `test_restart_repro_repeat` -> `test_repro_determinism_restart`

jo-basevi commented:
I've changed the marker names and experiment names in #159, and updated this PR to match.

chrisb13 commented:
Great, thanks @jo-basevi! So are we ready to merge / close this PR @dougiesquire?

dougiesquire reviewed (commented):
LGTM

jo-basevi commented:
The marker code updates #159 is now merged to main, so I think it's good to now merge this PR
